### ISSUE TITLE
Support for custom parameter type with `click.ParamType`

### ISSUE THREAD

**Issue by yashrathi-git** [tiangolo/typer#311]
Click have [click.ParamType](https://click.palletsprojects.com/en/8.0.x/parameters/#implementing-custom-types), for defining custom parameter types, but it doesn't work with typer (Example snippet below)

I want to use, my own formatting for datetime, ex: today %H:%M, %M:%H(automatically uses date as today). Typer already allows setting, [custom datetime formats](https://typer.tiangolo.com/tutorial/parameter-types/datetime/#custom-date-format) but it doesn't cover my use case.

This is example from [click docs](https://click.palletsprojects.com/en/8.0.x/parameters/#implementing-custom-types) that I tried using with **typer**.
```py
class BasedIntParamType(click.ParamType):
    name = "integer"

    def convert(self, value, param, ctx):
        if isinstance(value, int):
            return value

        try:
            if value[:2].lower() == "0x":
                return int(value[2:], 16)
            elif value[:1] == "0":
                return int(value, 8)
            return int(value, 10)
        except ValueError:
            self.fail(f"{value!r} is not a valid integer", param, ctx)

BASED_INT = BasedIntParamType()

def main(based_int: BASED_INT):
    print(based_int)
    
if __name__=="__main__":
    typer.run(main)
```
It gives this error:
```py
Traceback (most recent call last):
  File "/home/yashrathi/test.py", line 26, in <module>
    typer.run(main)
  File "/home/yashrathi/.local/lib/python3.9/site-packages/typer/main.py", line 859, in run
    app()
  File "/home/yashrathi/.local/lib/python3.9/site-packages/typer/main.py", line 214, in __call__
    return get_command(self)(*args, **kwargs)
  File "/home/yashrathi/.local/lib/python3.9/site-packages/typer/main.py", line 239, in get_command
    click_command = get_command_from_info(typer_instance.registered_commands[0])
  File "/home/yashrathi/.local/lib/python3.9/site-packages/typer/main.py", line 423, in get_command_from_info
    ) = get_params_convertors_ctx_param_name_from_function(command_info.callback)
  File "/home/yashrathi/.local/lib/python3.9/site-packages/typer/main.py", line 404, in get_params_convertors_ctx_param_name_from_function
    click_param, convertor = get_click_param(param)
  File "/home/yashrathi/.local/lib/python3.9/site-packages/typer/main.py", line 656, in get_click_param
    parameter_type = get_click_type(
  File "/home/yashrathi/.local/lib/python3.9/site-packages/typer/main.py", line 587, in get_click_type
    raise RuntimeError(f"Type not yet supported: {annotation}")  # pragma no cover
RuntimeError: Type not yet supported: <__main__.BasedIntParamType object at 0x7f7295822fd0>

```
`typer.main.get_click_type`, doesn't seem to be compatible with custom click type. But click knows how to handle  `click.ParamType`, `typer` doesn't need to convert it to click type.

Related Issue: #111

**Comment by dwreeves** [comment]
I really want something like this as I'm running into a similar issue transitioning some stuff over from Click to Typer.

One issue with the approach you're suggesting is that click param types are converted, so if you have something like:

```python
from typing import Optional
import click

class CustomParamType(click.ParamType):
    def convert(self, value, param, context) -> Optional[str]:
        if value is None:
            return None
        return str(value)

CUSTOM = CustomParamType()

def foo(x: CUSTOM):
    if x is not None:
        click.echo(x)
```

then your IDE doesn't know that `x` is actually an `Optional[str]` after conversion.

---

An alternate solution I'm a fan of is to add an attribute called `parameter_type` to:

- `ParameterInfo`
- `ArgumentInfo`
- `OptionInfo`
- `Argument`
- `Option`

then modify the function `get_click_param` to take the `parameter_type` from the parameter info class when `parameter_type is not None`.

So you could do something like this:

```python
import typer
from typing import Optional
import click

class CustomParamType(click.ParamType):
    def convert(self, value, param, context) -> Optional[str]:
        if value is None:
            return None
        return str(value)

CUSTOM = CustomParamType()

def foo(x: Optional[str] = typer.Argment(..., parameter_type=CUSTOM)):
    if x is not None:
        click.echo(x.lower())
```

^ With this approach, you may also want to check `isinstance(parameter_type, click.ParamType)` and warn or raise a custom error message when `issubclass(parameter_type, click.ParamType)`, otherwise the error messages could get cryptic.

**Comment by dwreeves** [comment]
I see there's some interest in this. I've opened a PR with the specification I outlined in my comment: https://github.com/tiangolo/typer/pull/337

**Comment by tiangolo** [comment]
Thanks! Support for this was added in https://github.com/tiangolo/typer/pull/583, the new docs are here: https://typer.tiangolo.com/tutorial/parameter-types/custom-types/, it's available in Typer `0.8.0`, just released. :tada:
